streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -98,16 +95,14 @@ public RecordCollectorImpl(final Producer<byte[], byte[]> producer, final String
                     @Override
                     public void onCompletion(final RecordMetadata metadata, final Exception exception) {
                         if (exception == null) {
-                            if (sendException != null) {
-                                return;
-                            }
                             final TopicPartition tp = new TopicPartition(metadata.topic(), metadata.partition());
                             offsets.put(tp, metadata.offset());
                         } else {
-                            if (sendException == null) {
-                                sendException = exception;
-                                log.error("{} Error sending record to topic {}. No more offsets will be recorded for this task and the exception will eventually be thrown", logPrefix, topic, exception);
-                            }
+                            log.error("{} Error sending record with key {} to topic {}, partition {}. " +
+                                    "No more offsets will be " +
+                                    "recorded for this task and the exception will eventually be thrown",
+                                logPrefix, key, topic, partition, exception);
+                            throw new StreamsException(String.format("%s exception caught when producing", logPrefix), exception);
```

Comment:
I think we cannot fix the issue, that error are detected late, as we want keep async pattern. I guess the problem is, that `checkException` is done within `send` at the beginning -- this confuses used as they assume the current send request fails.

Maybe we can do the check outside of `RecordCollectorImpl` ? Not sure -- might be hard to maintain. What we also can do, change the error message. Right now it only says "Error sending record to topic " -- maybe we can say "Aborting send record because a previous send returned an error"? I am also wondering, if the logged `topic` is correct -- should we not log `metadata.topic()` ?

We could also buffer up all sent records and include the record that causes the error in the log/exception -- to point out which record did cause the problem.